src/ppl/src/IOPlacer.cpp

```diff
@@ -1284,6 +1284,90 @@ void IOPlacer::findSlotsForTopLayer()
         top_layer_slots_.push_back({false, false, Point(x, y), top_grid_.layer, Edge::invalid});
       }
     }
+
+    filterObstructedSlotsForTopLayer();
+  }
+}
+
+void IOPlacer::filterObstructedSlotsForTopLayer()
+{
+  // Collect top_grid_ obstructions
+  std::vector<std::unique_ptr<odb::Rect>> obstructions;
+
+  // Get routing obstructions
+  for (odb::dbObstruction* obstruction : block_->getObstructions()) {
+    odb::dbBox* box = obstruction->getBBox();
+    if (box->getTechLayer()->getRoutingLevel() == top_grid_.layer) {
+      obstructions.push_back(std::make_unique<odb::Rect>());
+      box->getBox(*obstructions.back());
+    }
+  }
+
+  // Get already routed special nets
+  for (odb::dbNet* net : block_->getNets()) {
+    if (net->isSpecial()) {
+      for (odb::dbSWire* swire : net->getSWires()) {
+        for (odb::dbSBox* wire : swire->getWires()) {
+          if (!wire->isVia()) {
+            if (wire->getTechLayer()->getRoutingLevel() == top_grid_.layer) {
+              obstructions.push_back(std::make_unique<odb::Rect>());
+              wire->getBox(*obstructions.back());
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Get already placed pins
+  for (odb::dbBTerm* term : block_->getBTerms()) {
+    for (odb::dbBPin* pin : term->getBPins()) {
+      if (pin->getPlacementStatus().isFixed()) {
+        for (odb::dbBox* box : pin->getBoxes()) {
+          if (box->getTechLayer()->getRoutingLevel() == top_grid_.layer) {
+            obstructions.push_back(std::make_unique<odb::Rect>());
+            box->getBox(*obstructions.back());
+          }
+        }
+      }
+    }
+  }
+
+  // remove slots that go beyond the die boundary
+  odb::Rect die_area;
+  block_->getDieArea(die_area);
+  std::vector<Slot>::iterator it = top_layer_slots_.begin();
+  while (it != top_layer_slots_.end()) {
+    odb::Point& point = (*it).pos;
+    if (point.x() - top_grid_.width/2 < die_area.xMin()
+        || point.y() - top_grid_.height/2 < die_area.yMin()
+        || point.x() + top_grid_.width/2 > die_area.xMax()
+        || point.y() + top_grid_.height/2 > die_area.yMax()) {
+      // remove out of die pins
+      it = top_layer_slots_.erase(it);
```

Comment:
I think you can set the slot as blocked instead of erasing it from the vector. ioPlacer will not use any blocked slot.

Reply:
done, but I also had to modify getValidSlots to account for the different slots in the top layer versus normal slots.

Reply:
Looks good to me. I will let @QuantamHD approve the PR after his requests were implemented.